### ISSUE TITLE
The last 4 KiB of the address space are left out

### ISSUE THREAD

**Issue by axel-h** [seL4/seL4#571]
`tools/hardware/utils/memory.py` calculates 
```
    as_max = utils.align_down(config.addrspace_max, config.get_page_bits())
    ret.add(Region(0, as_max, None))
```
But since `config.addrspace_max` is of the form `2^n - 1` this means the last page is always  dropped. Printing gives me
`"Region [0x0..0xffffefff] (4294963200 byte)"`. With the changes of https://github.com/seL4/seL4/pull/556 the address space bit size is passed instead of a value that is reduced by one to avoid running into integer limits:
```
    as_size = 2**config.get_kernel_phys_addr_space_bits()
    ret.add(Region(0, as_size, None))   

````
Since integers in Python are not limited, that seem safe even for 64-bit architectures using the whole pyhs address space.

**Comment by lsf37** [comment]
You're right, this does not make much sense for `max`. If it is really a `max`, it should be the end of a page, not the beginning of a page.

In terms of python, no problem with keeping the size instead of the max here, this is not performance critical code in any way.

Main thing to look up is if the rest of the code is assuming a `max` or a `size` here. This is important for calculating overlap/adjacency etc, because they would then be off by one.

The code in seL4 in the end should be getting a `max`, so it might make sense to align things with that, but if the rest of the computations are easier as `size`, I don't mind much either way.

**Comment by kent-mcleod** [comment]
I had also noticed this before and have some unstaged changes for this:
```
diff --git a/tools/hardware/config.py b/tools/hardware/config.py
index 4a70bdb8..1bb7c709 100644
--- a/tools/hardware/config.py
+++ b/tools/hardware/config.py
@@ -25,6 +25,9 @@ class Config:
         ''' Get page size in bits for this arch '''
         return 12  # 4096-byte pages
 
+    def get_smallest_kernel_object_alignment(self) -> int:
+        return 4 # seL4_MinUntypedBits is 4 for all configurations
+
     def get_device_page_bits(self) -> int:
         ''' Get page size in bits for mapping devices for this arch '''
         return self.get_page_bits()
diff --git a/tools/hardware/utils/memory.py b/tools/hardware/utils/memory.py
index afbab9a0..3ee93348 100644
--- a/tools/hardware/utils/memory.py
+++ b/tools/hardware/utils/memory.py
@@ -98,7 +98,7 @@ def get_physical_memory(tree: FdtParser, config: Config) -> List[Region]:
 def get_addrspace_exclude(regions: List[Region], config: Config):
     ''' Returns a list of regions that represents the inverse of the given region list. '''
     ret = set()
-    as_max = utils.align_down(config.addrspace_max, config.get_page_bits())
+    as_max = utils.align_down(config.addrspace_max, config.get_smallest_kernel_object_alignment())
     ret.add(Region(0, as_max, None))
 
     for reg in regions:
```

This assumes that it's valid for `config.addrspace_max` is of the form `2^n - 1`.  As far as I'm aware, it was deliberate for the kernel to not give out an object that ends at the end of the address space.  This means that the kernel shouldn't be giving you the last page, but it could still give you smaller UT objects up to the boundary.

**Comment by axel-h** [comment]
If the kernel is not supposed to use an object at the end of the address space we should add this as a comment in the code. Also,I think it is cleaner and scales nicer if we put a reserved region at the end with exactly that explanation and create a region with the full address space here. Python seems to be able to handle arbitrary large integers, so there is no problem here. We can take care of this when creating C headers where this matters (assuming we have start/end pairs and not start/len pairs). I'm not sure about the YAML files, seem the also do not suffer from integer limits if Python parses them again.

**Comment by kent-mcleod** [comment]
Isn't it arbitrary whether we have a smaller address space, or a reserved region at the top? The reserved region at the top is because of the address space constraints, not because of firmware constraints, so it's simpler to have one address space region instead of an incorrect region and another reserved region that makes it correct?

**Comment by axel-h** [comment]
My thinking was to not restrict the general address space too much, but handle restrictions with regions then, where each region has an good explanation why it exists. Obviously, for very generic restrictions like in this case there is not much of a difference in the end. It seem beneficial to have the general address space bit size available and a max address that can be used separately. The ist just the constraint, that the max address can't exceed the address space bit size. But one should never have to guess the bit size from the max usable address then.
I might be a bit biased by my current PRs here...

**Comment by kent-mcleod** [comment]
> But since config.addrspace_max is of the form 2^n - 1 this means the last page is always dropped. 

This isn't true anymore for 64bit arches? It is still a case for 32bit.  Is addressing this on 32bit something that you are trying to achieve? I also don't think that it is an issue with the python scripts, but with the argument given to them.  I think the script is doing the right thing by treating the max address as exclusive.

**Comment by yyshen** [comment]
I vaguely remember the map_kernel_window for armv6 has some comments about the last page (I could be wrong), but it is probably too old to exist in this repo.

**Comment by lsf37** [comment]
This is now resolved in #1706
